### PR TITLE
Eng 1036 improve email validation

### PR DESCRIPTION
**Pull request by mailwilliams**
Issues:

- https://linear.app/metriport/issue/ENG-1036

### Dependencies

- Upstream: None
- Downstream: https://github.com/metriport/metriport-internal/pull/3255

### Description

Zod's email check isn't a fully comprehensive check (article from Zod's maintainer [here](https://colinhacks.com/essays/reasonable-email-regex)), so adding the validator package's isEmail accounts for more edge cases.

- Augmented zod's `string().email().safeParse` with validator package
  - Handles `!` in email edge case
  - Handles phone numbers being sent as emails (starting with `+1`) 
- Added more comprehensive testing 

### Testing

- Local
  - [x] Ran updated test suite
- Staging
  - [x] Updated email for test account

Check each PR.

### Release Plan

This update (given existing unit tests) should be backwards compatible with what we currently have, and if we are notified of new errors we can roll back the change

- [x] No dependencies between API and Infra that will cause errors during deployment
- [ ] Merge this

## Summary by CodeRabbit

* **New Features**
  * Broader email acceptance supporting many special characters in the local part and improved normalization.
  * New detection that treats phone-number-like inputs specially and returns a clear, specific error message.

* **Bug Fixes**
  * Prevents valid emails with special characters from being rejected or altered.

* **Documentation**
  * Expanded email docs with supported formats and explicit valid/invalid examples.

* **Chores**
  * Added a validated email library dependency to strengthen checks.

### CODE REVIEW

File: packages/api-sdk/src/medical/models/__tests__/demographics.test.ts [hunk 1]
```diff
@@ -31,6 +31,12 @@ describe("demographicsSchema", () => {
       "user123@metriport.com",
       "123user@metriport.com",
       "user@metriport.com.br",
+      "test!user@example.com",
+      "test+user@example.com",
+      "test_user@example.com",
+      "test-user@example.com",
+      "test.user@example.com",
+      "test_user+test@example.com",
     ];
```
[the remark this answers is not in the document]
Comment (mailwilliams) [reply packages/api-sdk/src/medical/models/__tests__/demographics.test.ts:40]: We have decided internally that if a user starts with +1, then it is most likely an error and we are returning an error. If we see that there are a lot of `+1actuallyvalid@...com` then we will address these edge cases later

File: packages/shared/src/domain/contact/__tests__/email.test.ts [hunk 2]
```diff
@@ -23,6 +24,54 @@ describe("Email Utility Functions", () => {
...
+      expect(isEmailValid("test|user@example.com")).toBe(true);
+      expect(isEmailValid("test}user@example.com")).toBe(true);
+      expect(isEmailValid("test~user@example.com")).toBe(true);
+    });
+
+    it("should return false for phone number format (+1 prefix)", () => {
+      expect(isEmailValid("+1234567890@example.com")).toBe(false);
+      expect(isEmailValid("+1-234-567-8900@example.com")).toBe(false);
+    });
+  });
```
[the remark this answers is not in the document]
Comment (mailwilliams) [reply packages/shared/src/domain/contact/__tests__/email.test.ts:57]: We have decided internally that if a user starts with +1, then it is most likely an error and we are returning an error. If we see that there are a lot of +1actuallyvalid@...com then we will address these edge cases later

File: packages/shared/src/domain/contact/email.ts [hunk 3]
```diff
@@ -11,9 +10,24 @@ export function isEmail(email: string): boolean {
...
+  if (isPhoneNumberFormat(email)) {
+    return false;
+  }
+
+  // Use our enhanced regex instead of Zod's potentially too-strict validation
+  return validator.isEmail(email, {
+    allow_utf8_local_part: true,
+    blacklisted_chars: "",
+  });
+}
```
[the remark this answers is not in the document]
Comment (mailwilliams) [reply packages/shared/src/domain/contact/email.ts]: We have decided internally that if a user starts with +1, then it is most likely an error and we are returning an error. If we see that there are a lot of +1actuallyvalid@...com then we will address these edge cases later

File: packages/shared/src/domain/contact/email.ts [hunk 4]
```diff
@@ -29,7 +43,14 @@
...
   const normalEmail = normalizeEmail(email);
-  if (!isEmailValid(normalEmail)) throw new Error("Invalid email.");
+  if (!isEmailValid(normalEmail)) {
+    if (isPhoneNumberFormat(email)) {
+      throw new Error(
+        "Invalid email: appears to be a phone number (starts with +1). Please enter a valid email address."
+      );
+    }
+    throw new Error("Invalid email.");
+  }
```
[the remark this answers is not in the document]
Comment (mailwilliams) [reply packages/shared/src/domain/contact/email.ts]: We have decided internally that if a user starts with +1, then it is most likely an error and we are returning an error. If we see that there are a lot of +1actuallyvalid@...com then we will address these edge cases later

File: packages/shared/src/domain/contact/email.ts [hunk 5]
```diff
@@ -11,9 +10,24 @@ export function isEmail(email: string): boolean {
...
+    allow_utf8_local_part: true,
+    blacklisted_chars: "",
+  });
+}
+
+/**
+ * Checks if an email appears to be a phone number (starts with +1)
+ */
+export function isEmailAPhoneNumber(email: string): boolean {
+  return email.trim().startsWith("+1");
```
Comment (keshavsaharia) [inline packages/shared/src/domain/contact/email.ts:32]: This is still the same as before - can we use `validator.isPhone(..., { countryCode: '+1' })` instead?

File: packages/shared/src/domain/contact/email.ts [hunk 6]
```diff
@@ -29,7 +43,14 @@ export function normalizeEmail(email: string): string {
  */
 export function normalizeEmailStrict(email: string): string {
   const normalEmail = normalizeEmail(email);
-  if (!isEmailValid(normalEmail)) throw new Error("Invalid email.");
+  if (!isEmailValid(normalEmail)) {
+    if (isEmailAPhoneNumber(email)) {
+      throw new Error(
```
Comment (keshavsaharia) [inline packages/shared/src/domain/contact/email.ts]: This should also be `BadRequestError`

File: packages/shared/src/domain/contact/__tests__/email.test.ts [hunk 7]
```diff
@@ -86,15 +145,25 @@ describe("email", () => {
...
     });
+
+    it("should handle emails with special characters", () => {
+      const input = "test!user@example.com";
+      expect(normalizeEmailNewSafe(input)).toBe("test!user@example.com");
+    });
+
+    it("should return undefined for phone number format", () => {
+      const input = "+1234567890@example.com";
+      expect(normalizeEmailNewSafe(input)).toBeUndefined();
```
Comment (keshavsaharia) [inline packages/shared/src/domain/contact/__tests__/email.test.ts:166]: Actually, I just read this test so it seems like the `+1` prefix is the expected behavior - let me know if that's the case.

File: packages/shared/src/domain/contact/email.ts (hunk 6 above)
Comment (mailwilliams) [reply packages/shared/src/domain/contact/email.ts]: Addressed

File: packages/shared/src/domain/contact/email.ts (hunk 5 above)
Comment (mailwilliams) [reply packages/shared/src/domain/contact/email.ts:32]: I updated the comment to explain that we internally decided an email starting with +1 is invalid. We can update the logic to remove this if we deem it necessary later on

File: packages/shared/src/domain/contact/__tests__/email.test.ts (hunk 7 above)
Comment (mailwilliams) [reply packages/shared/src/domain/contact/__tests__/email.test.ts:166]: Correct, this is expected behavior

**Review by keshavsaharia — APPROVED** [review]
